forecasttools/idata_w_dates_to_df.py

```diff
@@ -7,80 +7,226 @@
-                pl.date_range(
-                    start=start_date_as_dt,
-                    end=start_date_as_dt + pl.duration(days=interval_size - 1),
-                    interval="1d",
-                    closed="both",
-                    eager=True,
-                )
-                .to_numpy()
-                .astype("datetime64[ns]")
-            )
-            idata_group_with_dates = idata_group.assign_coords(
-                {dim_name: interval_dates}
-            )
-            setattr(
-                idata_w_dates,
-                group_name,
-                idata_group_with_dates,
+
+def add_time_coords_to_idata_dimensions(
+    idata: az.InferenceData,
+    groups: str | list[str],
+    variables: str | list[str],
+    dimensions: str | list[str],
+    start_date_iso: str | datetime,
+    time_step: timedelta,
+) -> az.InferenceData:
+    """
+    Modifies the time-based coordinates across
+    groups, variables, and dimensions. The
+    function checks if the group, variable,
+    and dimension exist, and then applies the
+    specified time-based coordinates.
+
+    Parameters
+    ----------
+    idata : az.InferenceData
+        The InferenceData object to modify.
+    groups : str | list[str]
+        A group or a list of groups to modify
+        (e.g., "posterior_predictive").
+    variables : str | list[str]
+        A variable or a list of variables
+        within the specified groups to modify.
+    dimensions : str | list[str]
+        A dimension or a list of dimensions
+        to modify for the variables.
+    start_date_iso : date | datetime
+        The start date for the time
+        coordinates as a datetime date or as
+        a datetime datetime.
+    time_step : timedelta
+        The time interval between each
+        coordinate (e.g., `timedelta(days=1)`).
+
+    Returns
+    -------
+    az.InferenceData
+        The modified InferenceData object with
+        updated time coordinates for the
+        specified groups, variables, and
+        dimensions.
+    """
+    inputs = [
+        (idata, az.InferenceData, "idata"),
+        (groups, (str, list), "groups"),
+        (variables, (str, list), "variables"),
+        (dimensions, (str, list), "dimensions"),
+    ]
+    for value, expected_type, param_name in inputs:
+        forecasttools.validate_input_type(
+            value=value, expected_type=expected_type, param_name=param_name
+        )
+    # if str, convert to list
+    groups = forecasttools.ensure_listlike(groups)
+    variables = forecasttools.ensure_listlike(variables)
+    dimensions = forecasttools.ensure_listlike(dimensions)
+    # check groups, variables, and dimensions
+    # all contain str vars
+    forecasttools.validate_iter_has_expected_types(groups, str, "groups")
+    forecasttools.validate_iter_has_expected_types(variables, str, "variables")
+    forecasttools.validate_iter_has_expected_types(
+        dimensions, str, "dimensions"
+    )
```

Comment:
Again, this feels like unnecessarily strong typing for a duck-y language, but I am less opposed here than I am above.